mini-projects/01-device-driver/laptop-interface/src/controller.rs

```diff
@@ -0,0 +1,516 @@
+#![allow(dead_code)]
+
+use crate::error::{ControllerError, ModuleError, SerialError};
+use crate::lib_serial_ffi::*;
+
+use std::ffi::CString;
+use std::sync::mpsc::{Receiver, Sender, TryRecvError};
+use std::thread::{self, JoinHandle};
+use std::time::Duration;
+use std::{default, str};
+
+use log::{self, error, info, trace};
+use regex::Regex;
+use std::collections::HashMap;
+
+pub struct ControllerManager {
+    controllers: Vec<Controller>,
+    input_receiver: Receiver<(u32, String)>,
+    output_sender: Sender<(u32, String)>,
+    pub join_handles: HashMap<u32, Option<JoinHandle<Result<(), ControllerError>>>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct ControllerState {
+    pub north_east: bool,
+    pub north_west: bool,
+    pub south_east: bool,
+    pub south_west: bool,
+    pub north: Option<bool>,
+    pub south: Option<bool>,
+    pub west: Option<bool>,
+    pub east: Option<bool>,
+
+}
+
+impl ControllerState {
+    /// Creates a new ControllerState from a bitmask.
+    /// Bit 3: Northwest button
+    /// Bit 2: Southwest button
+    /// Bit 1: Southeast button
+    /// Bit 0: Northeast button
+    pub fn from_bitmask(bitmask: u8) -> Self {
+        ControllerState {
+            north_east: (bitmask & 0b0001) != 0,
+            south_east: (bitmask & 0b0010) != 0,
+            south_west: (bitmask & 0b011) != 0,
+            north_west: (bitmask & 0b100) != 0,
```

Comment:
I'm convinced this bit mask analysis is wrong - I would think that since the northwest button is bit 3, you would just and the bit string with 0b00001000 to get the corresponding value